From 61fbe033507322775c8d020eaa7ee63db4dcdfb8 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Wed, 22 May 2019 17:01:09 +0200
Subject: [PATCH 250/386] net: phy: phylink: fix locking when updating SFP link
 status

Recent commit: cf4bca4034d7 ("net: phy: phylink: propagate MAC link
status change to the SFP layer") helped for the issue of missing
proper notification in SFP layer during NFS rootfs mount.

However the phylink does not allow to perform such update
without RTNL (a global lock for all changes to network configuration),
which is notified to user by a stack dump from ASSERT_RTNL().

Perform the update only if the RTNL lock is taken.

Change-Id: I7f51bcd2a6ef99b1456bf4dcf55a0bf25003dae7
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/9772
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/net/phy/phylink.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/phy/phylink.c b/drivers/net/phy/phylink.c
index 511d6c52e01f..7a216b0f5a1f 100644
--- a/drivers/net/phy/phylink.c
+++ b/drivers/net/phy/phylink.c
@@ -904,7 +904,7 @@ void phylink_mac_change(struct phylink *pl, bool up)
 
 #if IS_ENABLED(CONFIG_SFP)
 	/* Make sure the event is propagated to the SFP layer. */
-	if (pl->sfp_bus) {
+	if (pl->sfp_bus && rtnl_is_locked()) {
 		if (up)
 			sfp_link_up(pl->sfp_bus);
 		else
-- 
2.17.1

